Add nethealth checker

bernardjkim opened the pull request:
This PR adds the nethealth checker. [Design Doc](https://docs.google.com/document/d/1zJGGYLPWmNioanIpzuvtjZNffgFzJ-Ezkv3vP3kLKLM/edit#heading=h.208t1zijroe)

### Rationale
Satellite does not currently have a health check in place to verify network communication between peers. Blocked UDP traffic due to a firewall or other network issues may go undetected.

### Implementation
The [nethealth](https://github.com/gravitational/satellite/pull/113) service exposes a counter for the number of echo requests and timeouts from peers in a cluster. This checker pulls metrics from the nethealth service and verifies that the network communication between peers is functional. The timeout stats are recorded as a short time series data interval containing a user specified number of data points. If the packet loss percentage is above a specified threshold at each data point, the network will be considered unhealthy and will result in a failed check.

bernardjkim committed:
Add nethealth checker

knisbet commented:
So there is an issue with this approach that doesn't appear to have been caught on the design doc, when considering nethealth. By connecting to nethealth.monitoring.svc.cluster.local:9801/metrics, this uses the overlay network that may be blocked, and also accesses a random pod which is a single nodes perspective.

So each satellite agent should monitor the nethealth pod running on the same node, which provides only it's own perspective on connectivity to other nodes. That agent can then relay it's perspective over the peer-to-peer sync (presumably each node is running this check). So the way to do that, would be to instead do a DNS query for any.nethealth.monitoring.svc.cluster.local for the list of every POD providing the service, and then filtering to just the IP that belongs to the same node, and then scraping the metrics from the same node. 

Alternatively, the checker would need a kubernetes client, and to either watch or periodically query the pod objects, as the pod IP addresses can change if the pod is re-created, and then find the pod object with matching labels, that belongs to the same node as the checker. Going the DNS route is lighter weight overall.

r0mant commented:
@knisbet Wouldn't prometheus' service monitor exhibit the same behavior then? IIRC it also queries a service.

@bernardjkim I would go with the DNS approach Kevin mentioned above, but I think it makes sense  to query the local resolver (127.0.0.2) to ensure that traffic stays local. This means satellite will also need to find and cache information about the pod subnet of the node it's running on upon startup.

knisbet commented:
The subnet assigned to the host is available from within planet: /run/flannel/subnet.env
flannel writes this configuration when assigning a subnet. In theory it can change, but in practice it shouldn't.

@r0mant Prometheus uses the service to identify pods, but then I believe queries them individually. So you still get per pod/node results from prometheus, but configuration wise it's pointed at a service. Prometheus does exhibit similar limitations, the difference being that prometheus should report scraping failures to pods it's not able to contact, but if the overlay network is totally broken it may provide non intuitive results. The second component of which with the nethealth originally, was that on suspecting an overlay networking problem, the pod logs for nethealth could be observed to provide further insight. For partial interruptions, prometheus + alertmanager are able to report packet loss or totally interruption between any pair of nodes.

bernardjkim committed:
Update config

- Can now configure `NethealthIP` and `NethealthPort`. Expects planet to
provide correct ip and port for local nethealth service.
- Call `CheckAndSetDefaults` in constructor method.
- No longer need `node_name` if we are only collecting metrics from
local nethealth service.
- Add license header to `nethealth_test.go`.

bernardjkim commented:
So right now I'm planning on having planet be responsible for getting the IP address for the local nethealth service. The satellite nethealth checker will just query the metrics from the IP and port provided by planet.

We could do something like:
```golang
func getSubnet() (*net.IPNet, error) {
 	iface, err := net.InterfaceByName("cni0")
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 	addrs, err := iface.Addrs()
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
	if subnet, ok := addrs[0].(*net.IPNet); ok {
		return subnet, nil
	}
	return nil, trace.BadParameter("failed to get %s subnet", "cni0")
}
```

or:
```golang
func getSubnet() (*net.IPNet, error) {
	const envSubnetPath = "/run/flannel/subnet.env"
	const envSubnetKey = "FLANNEL_SUBNET"

	err := godotenv.Load(envSubnetPath)
	if err != nil {
		return nil, trace.NotFound("unable to find file %s", envSubnetPath)
	}

	subnetString := os.Getenv(envSubnetKey)
	if subnetString == "" {
		return nil, trace.NotFound("%s is not set", envSubnetKey)
	}

	_, subnet, err := net.ParseCIDR(subnetString)
	if err != nil {
		return nil, trace.Wrap(err)
	}

	return subnet, nil
}
```

Then check to see the local subnet contains any of the nethealth service ip addresses:
```golang
func getNethealthIP(host string) (net.IP, error) {
	subnet, err := getSubnet()
	if err != nil {
		return nil, trace.Wrap(err)
	}

	// Lookup all nethealth ip addresses.
	ips, err := net.LookupIP(host) // host = "any.nethealth.monitoring.svc.cluster.local"
	if err != nil {
		return nil, trace.Wrap(err)
	}

	// Return local nethealth ip.
	for _, ip := range ips {
		if subnet.Contains(ip) {
			return ip, nil
		}
	}
	return nil, trace.NotFound("unable to lookup local nethealth ip")
}
```

@knisbet, @r0mant what do you guys think?

r0mant commented:
The approach with using flannel's configuration to determine the subnet won't work if the flannel is disabled which happens if a user chooses to deploy their own cni plugin. I wonder if there's a way to do this without relying on flannel/interface. I'll think about it.

r0mant commented:
I mean, one idea which is similar somewhat to what Kevin mentioned above is to use Kubernetes API but instead of watching pods or filtering by labels, the agent can just query the correct pod by the advertise IP of the node (or query all and filter by advertise IP). So:

* Planet agents knows advertise address of the node its running on (or can figure out from /etc/container-environment).
* Then the agent can query the Kubernetes API for pods and find the one which runs on this host (should have .status.hostIP property same as the advertise IP).
* Then the agent makes query to this pod's IP.

One thing to keep in mind here is that on regular nodes "kubectl" does not have permissions to read pods so the agent will need to use k8s client with "kubelet" config (I think there should already be one).

bernardjkim committed:
Discover nethealth using k8s api

- NethealthConfig takes HostIP and KubeConfig.
- Query k8s to discover nethealth pod.

knisbet commented:
Yea, kubernetes client might be the way to go here if we want to support any potential CNI plugin. We don't necessarily have to though, since we only officially offer support for two plugins (flannel and wormhole).

bernardjkim commented:
Alright, made some changes in 1488378.
The nethealth checker now queries the kubernetes API for pods and filters using advertised IP.

a-palchikov commented on monitoring/nethealth.go:
```diff
@@ -226,11 +235,28 @@ func (c *nethealthChecker) setTimeoutSeries(name string, series []int64) error {
 
+	// Find nethealth pod with matching host ip address.
+	for _, pod := range pods.Items {
+		if !strings.Contains(pod.GetName(), nethealthName) {
```
You should list pods using specific label/field selectors to avoid this - otherwise you're running the risk of picking something arbitrary.

bernardjkim replied:
Addressed in f8d89e9.
Does this is implementation look correct?

bernardjkim committed:
Filter pods using labels

- Reorder config
- Filter pods using 'k8s-app' label

bernardjkim commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,293 @@
+	// number of echo timeouts received by the specific peer for a set interval.
+	timeoutStats *holster.TTLMap
+
+	// NethealthConfig contains caller specified nethealth checker configuration
```
f8d89e9

bernardjkim commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,293 @@
+// verifyNethealth verifies that the network communication is healthy for the
+// nodes specified by the provided list of names.
+func (c *nethealthChecker) verifyNethealth(names []string) error {
+	var errors []error
```
Should we be completely avoiding `NewAggregate` for now? Currently using `NewAggregate` in `CheckAndSetDefaults`, `updateStats`, and `verifyNethealth`. Should these all be replaced to fail fast and return individual errors? 

If we aggregate the errors for `verifyNethealth` we can return a single probe indicating all the nodes that it is unable to communicate with:
```bsh
* dev-bk-gravity-node-0 / 10.128.0.70 / node
            Status:     degraded
            [×]         failed nethealth test (network communication failure with 10.128.0.96, network communication failure with 10.128.0.97)
        * dev-bk-gravity-node-1 / 10.128.0.97 / node
            Status:     degraded
            [×]         failed nethealth test (network communication failure with 10.128.0.70, network communication failure with 10.128.0.96)
        * dev-bk-gravity-node-2 / 10.128.0.96 / node
            Status:     degraded
            [×]         failed nethealth test (network communication failure with 10.128.0.70, network communication failure with 10.128.0.97)
```

without aggregating errors we can return multiple probes, one for each peer:
```bsh
        * dev-bk-gravity-node-0 / 10.128.0.70 / node
            Status:     degraded
            [×]         nethealth-checker (network communication failure with 10.128.0.96)
            [×]         nethealth-checker (network communication failure with 10.128.0.97)
        * dev-bk-gravity-node-1 / 10.128.0.97 / node
            Status:     degraded
            [×]         nethealth-checker (network communication failure with 10.128.0.70)
            [×]         nethealth-checker (network communication failure with 10.128.0.96)
        * dev-bk-gravity-node-2 / 10.128.0.96 / node
            Status:     degraded
            [×]         nethealth-checker (network communication failure with 10.128.0.70)
            [×]         nethealth-checker (network communication failure with 10.128.0.97)
```

The second output looks a bit easier to read. Is this format more consistent with how were reporting failed probes?

bernardjkim commented:
With the current implementation, the probes do not provide much insight into why the nethealth checker is failing.

Example:
If we have a 3 node cluster (node-0, node-1, node2) and add a firewall rule to node-1 to drop udp traffic from node-0 we will end up with:
```bsh
* dev-bk-gravity-node-0 / 10.128.0.70 / node
    Status:     degraded
    [×]         nethealth-checker (network communication failure with 10.128.0.97)
* dev-bk-gravity-node-1 / 10.128.0.97 / node
    Status:     degraded
    [×]         nethealth-checker (network communication failure with 10.128.0.70)
* dev-bk-gravity-node-2 / 10.128.0.96 / node
    Status:     healthy
```
The failed probes indicate that node-0 and node-1 are failing to communicate, but doesn't really have any indication why.

Should we add additional steps to this checker? Once we detect that the overlay network is unhealthy, should we run additional tests to diagnose common problems with the network. For example, checking iptables? Not sure what other common problems users have. 

@r0mant, @knisbet Is this worth looking into or outside the scope of this feature?

r0mant commented:
I think this is good for now. There is a clear indication that there is a network communication problem b/w these 2 nodes which is a good kick-start for troubleshooting - and there may be myriads of reasons. It would certainly be helpful to be able to detect the most common ones, but I don't think we should be spending time on this right now.

r0mant commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,321 @@
+		}
+
+		// Keep only the last `seriesCapacity` number of data points.
+		if len(series) >= c.SeriesCapacity && len(series) > 0 {
```
Nit: I think `c.SeriesCapacity` is always greater than 0 so the second `&& len(series) > 0` check isn't necessary here.

r0mant commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,321 @@
+// the provided metrics. Returns a list containing the updated keys.
+func (c *nethealthChecker) updateStats(metrics []*dto.Metric) (updated []string, err error) {
+	var errors []error
+	for _, metric := range metrics {
```
So what are "metrics" here? Is it a single datapoint per peer representing whether the last nethealth check was success/failure? Or are these multiple binary (1/0) datapoints for each peer? Or number of timeouts for each peer? Could you describe the data model somewhere in the comments please.

r0mant commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,321 @@
+
+	// Increase in number of timeouts at each data point indicates network issue.
+	for i := 0; i < c.SeriesCapacity-1; i++ {
+		if series[i] >= series[i+1] {
```
Trying to understand this algo - this kinda goes to my question about the data model above, what does `series[i]` represent? Number of timeouts during previous interval?

If so, won't this algorithm fail to detect the following scenario for example:

`5,6,5,6,5,6,5,6,..`

i.e. the number of timeouts decreases and the checker will think it's healthy. Unless I'm missing something (will prob be more clear when you document the data model).

r0mant commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,321 @@
+			return fmt.Sprintf("http://%s:%d", pod.Status.PodIP, c.NethealthPort), nil
+		}
+	}
+	return "", trace.NotFound("unable to find local nethealth pod")
```
Scenario where nethealth pod can't be found should be handled gracefully. Think about the upgrades - a new planet agent comes up but monitoring app still hasn't been upgraded and as such there's no nethealth pod running yet.

I would handle this not found error on the caller side gracefully and maybe just log it but not fail the checker.

bernardjkim commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,321 @@
+	sync.Mutex
+	// timeoutStats maps a peer to a time series data interval containing the
+	// number of echo timeouts received by the specific peer for the set interval.
+	timeoutStats *holster.TTLMap
```
I saw that previous usage of `gravitational/ttlmap` was replaced in https://github.com/gravitational/satellite/pull/163. I haven't really looked at the differences, but I just figured we're moving from `ttlmap` to `holster`.

bernardjkim committed:
Edit NethealthConfig

- HostIP -> AdvertiseIP
- Now saves a KubeConfig pointer

bernardjkim committed:
Add nethealthMetric struct

- `parseMetric` now parses the input into an array of `nethealthMetric`.
This should help clarify the logic.

bernardjkim committed:
Remove error aggregates

- `updateStatus` and `verifyNethealth` no longer return error aggregate.
- `verifyNethealth` will report individual probes for each peer.
- Errors no longer trigger failed probe, they are all just logged as
errors.

bernardjkim committed:
Update nethealth_test.go

bernardjkim committed:
Clarify `isHealthy` logic

r0mant reviewed (commented):
I think the algo needs some more tweaking to be more reliable/predictable, see my comment.

bernardjkim committed:
Report k8s & nethealth communication failures

- Will report failed probe if unable to retrieve pods from k8s.
- Will report failed probe if unable to fetch metrics from nethealth
after addr has been found.
- Increase timeout stats TTL to 1 day.

bernardjkim committed:
Update algo to determine network health

- Nethealth checker now keeps track of the packet loss as a percentage
for each peer. Overlay network will be considered unhealthy if
the percentage is above the threshold throughout the interval.

bernardjkim commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,340 @@
+// Implements health.Checker
+func (c *nethealthChecker) Check(ctx context.Context, reporter health.Reporter) {
+	if err := c.check(ctx, reporter); err != nil {
+		log.WithError(err).Error("Unable to verify nethealth.")
```
c234ed6
Failure to get k8s pods or fetch metrics from nethealth pod will now report failed probes as well.

bernardjkim commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,340 @@
+	// this peer have been timing consistently throughout this interval. This
+	// should indicate that there is a network issue.
+	for i := 0; i < c.SeriesCapacity-1; i++ {
+		if series[i] >= series[i+1] {
```
6e2409c
Determining nethealth using packet loss percentage should definitely be more reliable. This commit adds a default 20% packet loss threshold and a 5 minute interval. The interval is used along the the agent's status update timeout to calculate the number of data points to store in a series.

bernardjkim committed:
Update nethealth test

bernardjkim committed:
Fix isHealthy algo

Was previously calculating packet loss percentage using the current
running total of requests and timeouts. We only want to look at the
increaste in requests and timeouts between each check.

r0mant commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,411 @@
+	NethealthConfig
+	// seriesCapacity specifies the number of data points to store in a series.
+	seriesCapacity int
+	// lock access to timeoutStats
```
timeoutStats -> peerStats?

r0mant commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,411 @@
+		log.WithError(err).Error("Unable to verify nethealth.")
+	}
+
+	if probes.NumProbes() != 0 {
```
To be honest, instead of returning probes here, I would still pass reporter to check() function like it was before, but just made sure that check() returns "nil" in case nethealth pod wasn't found. This way it would plug seamlessly into the old logic and it would be cleaner because probes would only be created in a single place. For example:

```go
func (c *nethealthChecker) Check(ctx context.Context, reporter health.Reporter) {
  probes, err := c.check(ctx)
  if err != nil {
    // add error probe
  }
  // add success probe
}

func (c *nethealthChecker) check(ctx context.Context, reporter health.Reporter) error {
  addr, err := c.getNethealthAddr(pods)
  if err != nil && !trace.IsNotFound(err) {
    return trace.Wrap(err) // we got unexpected error, maybe networking-related, will add error probe above
  }
  if trace.IsNotFound(err) {
    log.Debug("Nethealth pod not found.")
    return nil // pod wasn't found, log and treat gracefully
  }
  //...
}

func (c *nethealthChecker) getNethealthAddr() (string, error) {
  pods, err := c.Client.CoreV1().Pods(nethealthNamespace).List(metav1.ListOptions{})
  if err != nil {
    return "", rigging.ConvertError(err) // this will convert error to a proper trace error, e.g. trace.NotFound
  }
  // ...
}
```

bernardjkim replied:
Ah ok, reverting back to previous structure. 
Currently unable to import a compatible version of `rigging`. Opened a PR to bump k8s and set a trace version https://github.com/gravitational/rigging/pull/75

r0mant replied:
Hmm, since we need this in 5.5 I'm not sure we can safely bump Kubernetes version in rigging and revendor here. Might run into incompatibility issues - 5.5 has Kubernetes 1.13.

r0mant replied:
@bernardjkim Btw on that note - I have just noticed that this PR is open to master. Shouldn't it be pointing to version/5.5.x instead? Because we're essentially backporting all this to 5.5 so all code changes should go to version/5.5.x branches of respective codebases (satellite, planet, etc.)

r0mant commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,411 @@
+
+func (c *nethealthChecker) check(ctx context.Context) (health.Reporter, error) {
+	reporter := new(health.Probes)
+	pods, err := c.Client.CoreV1().Pods(nethealthNamespace).List(metav1.ListOptions{})
```
Like in my comment above, this should be encapsulated into getNethealthAddr.

r0mant commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,411 @@
+	reporter := new(health.Probes)
+	pods, err := c.Client.CoreV1().Pods(nethealthNamespace).List(metav1.ListOptions{})
+	if err != nil {
+		reporter.Add(NewProbeFromErr(c.Name(), "failed to retrieve k8s pods", err))
```
And these will go away as we'll only be adding probes in the top-level Check method.

r0mant commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,411 @@
+		return reporter, trace.Wrap(err, "failed to get local nethealth address")
+	}
+
+	mf, err := fetchMetrics(ctx, addr)
```
Nit: What's "mf"?  Can we pick a better name for this variable?

r0mant commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,411 @@
+
+	resp, err := client.Get(ctx, client.Endpoint("metrics"), url.Values{})
+	if err != nil {
+		return nil, trace.ConvertSystemError(err)
```
Should be just trace.Wrap() I think, ConvertSystemError is usually used to convert Go errors from system methods like open file, etc.

r0mant commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,411 @@
+// updateStats updates netStats with new data collected in the provided
+// MetricFamily.
+func (c *nethealthChecker) updateStats(echoRequests, echoTimeouts map[string]float64) (updated []string, err error) {
+	if len(echoRequests) != len(echoTimeouts) {
```
Do these lengths have to match? What do these maps map from->to?

bernardjkim replied:
They should match. The metrics should come in a pair (requestTotal, timeoutTotal). I've just wrapped both into a `networkData` struct and have `parseMetrics` return `networkData`

r0mant commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,411 @@
+		// It should not be possible for the timeout counter to have increased
+		// more than the request counter. Log and ignore this situation.
+		if timeoutInc > requestInc {
+			log.WithField("request inc", requestInc).
```
Nit: We usually make log fields w/o spaces, "request-increase", "timeout-increase".

r0mant commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,411 @@
+		// Request counter should be constantly increasing, if this is not the
+		// case, nethealth pod most likely reset the counter. Do no update
+		// packet loss percentages this round.
+		if requestInc <= 0 {
```
Hm, so if nethealth pod restarts, this will keep being < 0 until the previously reached number of total requests is reached again, or am I missing anything? Should we reset the peer stats in this case also?

bernardjkim replied:
Ah you're right. The stored previous totals should be updated here.

r0mant commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,411 @@
+		if len(data.packetLossPercentages) >= c.seriesCapacity {
+			data.packetLossPercentages = data.packetLossPercentages[1:]
+		}
+		packetLoss := timeoutInc / requestInc
```
Interesting, so you're recording the packet loss after each interval and then compare each of them with the threshold below. Do you know how many datapoints nethealth produces b/w the checker runs? I'm just thinking, if it checks it once every 30 seconds, we may have not enough "increases" here and the packet loss will always be either close to 0 or to a 100. Have you considered to instead just accumulating these increases and then calculating packet loss once in "isHealthy" function - so we'd get the packet loss for the entire interval.

bernardjkim replied:
After thinking about this some more, I think calculating the overall packet loss once for the interval may give some undesired results.

Example:
Two nodes are unable to communicate on the overlay network for some time and end up with 100% packet loss for the entire collection interval. When the overlay network issue is solved and the two nodes are able to communicate, the checker will continue to be fail until the overall packet loss percentage has decreased below the threshold. If starting from 100% this will probably require no packet loss for the entire collection interval.

r0mant commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,411 @@
+			return label.GetValue(), nil
+		}
+	}
+	return "", trace.NotFound("unable to find required peer label")
```
Error could be a bit better, like which label?

bernardjkim replied:
Modified to print out `peer_name` label.

a-palchikov commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,411 @@
+		c.NetStatsInterval = defaultNetStatsInterval
+	}
+	if c.KubeConfig == nil {
+		errors = append(errors, trace.BadParameter("kubernetes access config must be provided"))
```
nit: move the test before the defaults are set.

bernardjkim committed:
Fix godoc, formatting, var names, etc...

- timeoutStats -> peerStats
- mf -> metricFamilies
- ConvertSystemError -> Wrap
- "request inc" -> "request-inc"
- log "peer_name" label
- Move tests before defaults are set.

bernardjkim committed:
Add reporter to `check`

bernardjkim committed:
Add ConvertError util function

bernardjkim committed:
Calculate health using packet loss over interval

bernardjkim committed:
Calculate individual packet loss percentages

bernardjkim committed:
Fix lint errors

- netStats avoid passing lock by value
- Remove unused func

a-palchikov commented on monitoring/nethealth_test.go:
```diff
@@ -0,0 +1,211 @@
+			storedData: peerData{
+				requestInc: []float64{10, 20, 30, 40, 50},
+				timeoutInc: []float64{10, 20, 30, 40, 50},
+			},
```
You don't exercise the length difference test between requests/timeouts which is handled in code.

bernardjkim replied:
1c9ec8b
Added some test cases for parseMetrics.

bernardjkim commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,448 @@
+
+// getNethealthAddr returns the address of the local nethealth pod.
+func (c *nethealthChecker) getNethealthAddr() (addr string, err error) {
+	pods, err := c.Client.CoreV1().Pods(nethealthNamespace).List(metav1.ListOptions{})
```
Oh, did not realize we could filter the pods with `ListOptions`, but that makes sense that functionality would be provided by k8s api. I'll try to be more thorough when learning to use a new api.

bernardjkim commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,448 @@
+	// Request counter should be constantly increasing, if this is not the
+	// case, nethealth pod most likely reset the counter. Update prevTotals,
+	// but do no update requestInc/timeoutInc.
+	if requestInc <= 0 {
```
Yes, `timeoutInc` could also be negative if the nethealth pod restarts. I'll add an additional check.

bernardjkim committed:
Fix godoc and typos

bernardjkim committed:
Modify netStats

- netStats constructor now accepts a packetLossCapacity. This value is
used to initialize packetLoss at a specific capacity.
- peerData now stored the calculated packet loss instead of
request/timeout counters.

bernardjkim committed:
Query pods with ListOptions and Selectors

bernardjkim committed:
Log metrics string in case we fail to parse

bernardjkim committed:
Update nethealth_test.go

- Remove non-essential information.
- Add test cases for parseMetrics.

bernardjkim commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,448 @@
+	}
+
+	// Checker has not collected enough data yet to check network health.
+	if len(storedData.requestInc) < c.seriesCapacity {
```
2365796
Modified `netStats` and can now be initialized with a `packetLossCapacity`. `packetLoss` will be initialized with a constant capacity.

bernardjkim commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,448 @@
+	// Each peer should have a request counter and timeout counter pair.
+	// If lengths are unequal, we are missing data.
+	if len(echoRequests) != len(echoTimeouts) {
+		return nil, trace.BadParameter("recieved %d timeout counters and %d request counters",
```
2f089b7
Moved parsing functionality into fetchMetrics. If we fail to parse metrics we will now log the metrics in text format received from nethealth. Also added an example of expected nethealth metrics. 

We will only run into this situation if there is an bug in nethealth.

a-palchikov commented on monitoring/nethealth.go:
```diff
@@ -203,17 +204,19 @@ func (c *nethealthChecker) updatePeer(peer string, incomingData networkData) err
 	}
 
-	// Calcuate counter increase since last check and replace previous total
+	// Calcluate counter delta since last check and replace total.
```
Cal**cul**ate :-)

a-palchikov commented on monitoring/nethealth.go:
```diff
@@ -203,17 +204,19 @@ func (c *nethealthChecker) updatePeer(peer string, incomingData networkData) err
-		log.Warn("Request counter did not increase. Nethealth pod may have restarted.")
+	// Request counter should be strictly increasing and timeout counter should
+	// be monotonically increasing. If this is not the case, nethealth pod most
+	// likely restart and reset the counters.
```
restart**ed** and reset the counters

bernardjkim committed:
Bump trace to 1.1.10

bernardjkim committed:
Clarify errors and godocs

bernardjkim committed:
Fix nethealth bugs and error handling

- Parsing error no longer affects status
- Additional error logging
- Fix sequence overflow bug
- Fix getNethealthAddr by removing field selector
- Close expired client connections in timedrift checker

bernardjkim committed:
Bump nethealth build base image 18.10 -> 19.10
